lgc/patch/gfx9/llpcNggPrimShader.cpp

```diff
@@ -3316,15 +3316,22 @@ Function* NggPrimShader::MutateEsToVariant(
 
     auto savedInsertPos = m_pBuilder->saveIP();
 
-    BasicBlock* pRetBlock = &pEsEntryVariant->back();
-    m_pBuilder->SetInsertPoint(pRetBlock);
+    // Find the return block and remove old return instruction
+    BasicBlock* pRetBlock = nullptr;
+    for (BasicBlock& block : *pEsEntryVariant)
+    {
+        auto pRetInst = dyn_cast<ReturnInst>(block.getTerminator());
+        if (pRetInst != nullptr)
+        {
+            pRetInst->dropAllReferences();
+            pRetInst->eraseFromParent();
 
-    // Remove old "return" instruction
-    assert(isa<ReturnInst>(pRetBlock->getTerminator()));
-    ReturnInst* pRetInst = cast<ReturnInst>(pRetBlock->getTerminator());
+            pRetBlock = &block;
+            break;
```

Comment:
And there cannot be multiple return blocks?

Reply:
Only one return block. Because we process this in `SpirvLowerGlobal::LowerOutput`. We create only one return block and add necessary "branch" instructions (jump to this return block) if there are multiple "return" in GLSL main function. Therefore, outputs always stay in this return block safely.